### PR TITLE
[TD-2260] Pre Chat Collection Support 

### PR DESCRIPTION
**Pull request by Sathyan-Elangovan**
**Feature Description:**
If user enables `Lead Collection` in the Dashboard, In the Chat widget when user clicked` Login AS Visitor`, instead of creating random user, widget will show the `User Lead Collection` Screen and after collecting those details, 
widget will create new User based on those details.

once has been created & registered, chat widget launch
1. conversation list screen if user has multiple conversation.
2. particular previous chat screen if user has only one conversation.
3. create a new conversation & show conversation screen if user has no previous conversation.

### CODE REVIEW

File: kommunicate/src/main/java/io/kommunicate/Kommunicate.java [hunk 1]
```diff
@@ -158,18 +160,155 @@ public void onFailure(RegistrationResponse registrationResponse, Exception excep
...
+        } catch (ClassNotFoundException e) {
+            throw new KmException(e.getMessage());
+        }
+    }
+
     public static void loginAsVisitor(Context context, KMLoginHandler handler) {
         login(context, getVisitor(), handler);
     }
 
+    public static void loginAsVisitorV2(final Context context, final KmCallback callback) {
```
Comment (amntoppo) [inline kommunicate/src/main/java/io/kommunicate/Kommunicate.java:203]: rather than using "loginAsVisitorV2", we can use "loginWithPrechat"

Comment (Sathyan-Elangovan) [reply kommunicate/src/main/java/io/kommunicate/Kommunicate.java:203]: Yeah, this is for our reference not the final method name.Once the code testing & review is will update the name as **launchConversationWithPreChat()**

**Review by amntoppo — APPROVED** [review]

File: app/src/main/java/kommunicate/io/sample/MainActivity.java [hunk 2]
```diff
@@ -101,18 +101,32 @@ public void onClick(View v) {
                 progressDialog.setCancelable(false);
                 progressDialog.show();
                 Kommunicate.init(MainActivity.this, APP_ID);
-                Kommunicate.loginAsVisitor(MainActivity.this, new KMLoginHandler() {
+//                Kommunicate.loginAsVisitor(MainActivity.this, new KMLoginHandler() {
```
Comment (mukeshthawani) [inline app/src/main/java/kommunicate/io/sample/MainActivity.java:104]: Remove this commented code.

File: kommunicate/src/main/java/io/kommunicate/Kommunicate.java [hunk 3]
```diff
@@ -158,18 +161,164 @@ public void onFailure(RegistrationResponse registrationResponse, Exception excep
...
+            @Override
+            public void onFailure(Object error) {
+                Utils.printLog(context, TAG, "Failed to fetch AppSetting");
+                loginUserWithKmCallBack(context, kmUser, callback);
+
+            }
+        }).execute();
+    }
+
+    // To Login the user from Lead Collection Details and Launch the Chat directly
```
Comment (mukeshthawani) [inline kommunicate/src/main/java/io/kommunicate/Kommunicate.java:265]: Can you also add method docs which includes argument details?

File: app/src/main/java/kommunicate/io/sample/MainActivity.java (hunk 2 above)
Comment (Sathyan-Elangovan) [reply app/src/main/java/kommunicate/io/sample/MainActivity.java]: @mukeshthawani removed.

File: kommunicate/src/main/java/io/kommunicate/Kommunicate.java (hunk 3 above)
Comment (Sathyan-Elangovan) [reply kommunicate/src/main/java/io/kommunicate/Kommunicate.java]: @mukeshthawani  done

**Review by mukeshthawani — APPROVED** [review]

**Pull request merged**